Subscribers: Add sharing network icons and links

ramonjd opened the pull request:
Fixes https://github.com/Automattic/wp-calypso/issues/97973


## Proposed Changes

This PR:

- adds social media and [Web Share API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Share_API)  buttons to the share site launchpad model.
- replaces HTML with common Gutenberg component patterns, e.g., button, flex containers
- ensures consistency with spacing and headings with the other modals, e.g., Import existing subscribers
- fixes the overflow in very looooong domain names

### Before


<img width="670" alt="Screenshot 2025-02-14 at 4 48 55 pm" src="https://github.com/user-attachments/assets/c3e20f29-5331-416d-a468-bb1388399059" />

### After

![2025-03-12 11 25 35](https://github.com/user-attachments/assets/09d3197a-90db-4ba9-90f6-be6bc3f66ebc)

<img width="700" alt="Screenshot 2025-03-12 at 11 17 02 am" src="https://github.com/user-attachments/assets/af949ed0-b90c-4ebf-a30b-d7c43092184e" />



## Why are these changes being made?

As noted in https://github.com/Automattic/wp-calypso/issues/97973, the current screen is confusing:

- It urges users to copy the link and _"...head over to your site and share it with the world"_. How is pasting my site link on my own site sharing it with the world?
- It offers no alternatives to share the link with the world.
- It's inconsistent with other modals. Let's just use the default padding and title attribute.

## Testing Instructions

Go to `/subscribers/[YOUR_SITE_SUBDOMAIN].wordpress.com` and click on "Share your site"

If you've already shared your site the options might be missing. 

If you need to reset the launch pad options, you can delete the `launchpad_checklist_tasks_statuses` blog option. E.g., 

```
echo switch_to_blog(get_blog_id_from_url('[YOUR_SITE_NAME].wordpress.com'));
echo delete_option('launchpad_checklist_tasks_statuses');
```

Click on every action button to ensure the links work.

Ensure that clicking on any share link triggers the launchpad status, e.g., Share your site will be checked off


<img width="635" alt="Screenshot 2025-03-10 at 2 11 58 pm" src="https://github.com/user-attachments/assets/417d09f8-c25b-4fe1-9a42-0bd05263a92a" />


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

ramonjd committed:
First commit only. Add sharing network icons and links

ramonjd committed:
Remove unnecessary sub text

ramonjd committed:
This commit adds the web share API. It also completes the details for each social network.

ramonjd committed:
$white doesn't exist and precommit checks whine about white. So #fff it is.

tellthemachines reviewed (commented):
Just did a quick test and this seems to be working as expected.

One thing I noticed in trying to share to WP is that it opens the editor and adds an embed block, that then says the content can't be embedded:

<img width="653" alt="Screenshot 2025-02-13 at 11 20 55 am" src="https://github.com/user-attachments/assets/89e7168b-c07e-4be3-82c5-96122f497f68" />

Looking at the block code, that message appears if there's no valid preview: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/embed/edit.js#L99
Is there anything we can do ensure there is a valid preview for the site? Otherwise, might there be a more appropriate block to use here? 
I'm not sure that this is in scope for this PR! But might be worth investigating separately if not.

ramonjd commented:
> Is there anything we can do ensure there is a valid preview for the site? Otherwise, might there be a more appropriate block to use here?

Thanks for testing! 

I can reproduce. Looks like the embed can only be used for single pages 🤔 

Maybe I'll remove the WP button for now. Not sure, if there's a another link URL.

ramonjd committed:
Remove WordPress since host URLs aren't embeddable. Updated icon type to conform with SocialLogo component.

ramonjd commented:
> I haven't yet touched the launchpad settings or included any tracking on the buttons. What do folks recommend here? Should the launch task be fulfilled by clicking on any share button? Should clicking on individual social networks be tracked?

Maybe it's something we can implement after the UX improvement if required?

ramonjd commented:
Note to self: review usage of network icons in light of https://github.com/Automattic/wp-calypso/pull/100807

ramonjd committed:
Refactors the flex containers. Tracks launchpad clicks for all share types. Removes dead code.

ramonjd committed:
Expected "1.2" to be one of "0.75, 0.875, 1, 1.25,  scales/font-sizes

andrewserong reviewed (commented):
Nice work, overall this is testing well for the services where I have an account 👍 

<img width="687" alt="image" src="https://github.com/user-attachments/assets/d110229f-df9c-4001-ae47-823bd73ac30c" />

Code-wise it's all looking good, just noticed an encoding issue with the Telegram share, so I've left a comment. If designers are happy with how it looks it's otherwise feeling close to land to me!

andrewserong commented on packages/launchpad/src/action-components/share-site-modal/index.tsx:
```diff
@@ -14,11 +20,84 @@ interface ShareSiteModalProps {
+			label: __( 'LinkedIn', 'launchpad' ),
+		},
+		{
+			href: `https://t.me/share/url?url=${ encodeURIComponent(
+				encodedSiteUrl
+			) }&text=${ encodedText }`,
+			className: 'share-site-modal__modal-share-link',
+			title: __( 'Share on Telegram', 'launchpad' ),
+			icon: 'telegram',
+			label: __( 'Telegram', 'launchpad' ),
+		},
```
This appears to be double-encoding for me:

<img width="475" alt="image" src="https://github.com/user-attachments/assets/b4942eaf-0abd-4359-8e1b-2958190963a2" />

If I manually update the URL then it looks okay:

<img width="498" alt="image" src="https://github.com/user-attachments/assets/c869cd8d-63b7-4290-8877-7cabc4d74cc5" />

Do we need the additional `encodeURIComponent` call?

ramonjd replied:
No we don't! That was a last minute copy pasta. Thanks for catch 🙇🏻

ramonjd replied:
Fixed!

andrewserong replied:
That's testing well for me, now 👍

ramonjd commented on packages/launchpad/src/action-components/share-site-modal/index.tsx:
```diff
@@ -14,11 +20,84 @@ interface ShareSiteModalProps {
+			label: __( 'LinkedIn', 'launchpad' ),
+		},
+		{
+			href: `https://t.me/share/url?url=${ encodeURIComponent(
+				encodedSiteUrl
+			) }&text=${ encodedText }`,
```
```suggestion
			href: `https://t.me/share/url?url=${ encodedSiteUrl }&text=${ encodedText }`,
```

ramonjd committed:
Update packages/launchpad/src/action-components/share-site-modal/index.tsx

jasmussen commented:
Nice, thank you for pinging the design group! To be sure I understand: this list of links is new, so you're looking for design feedback on how this visually looks before shipping?

It looks pretty good:

![Screenshot 2025-03-11 at 09 47 34](https://github.com/user-attachments/assets/240d1eb5-0907-4087-83e5-378e45bcb862)

But it can look better. How about:

![Screenshot 2025-03-11 at 09 56 59](https://github.com/user-attachments/assets/1da2e0a9-4d3c-4255-8613-6c36027f8a2e)

This removes the backgrounds, and the circles around the sharing buttons.

It also removes the "Share" button, which I'm unsure what does since you have additional sharing buttons below (but let me know if I'm missing nuance here).

A longer term look at this, I wouldn't use custom componentry at all, here it is with core components/patterns:

![Screenshot 2025-03-11 at 09 58 55](https://github.com/user-attachments/assets/8be9eecb-46ea-4a69-ab32-9e8f74796201)

_In both of the above, I'm not suggesting the social icons be replaced with cogs, I just didn't immediately find sources for the mockup work, hopefully the point gets across._

The future iteration isn't necessarily a blocker or urgent, but I like to always look ahead, it helps us ask questions for the near term.

ramonjd commented:
Thanks @jasmussen !! 🙇🏻 

I'll update with those changes.

> It also removes the "Share" button, which I'm unsure what does since you have additional sharing buttons below (but let me know if I'm missing nuance here).

If you look at the screenshot in the PR description it shows that this button triggers the [browser's web share API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Share_API). I think it's a neat feature, especially for mobile users. Maybe it could be better labelled (More sharing options...) or moved down near the socials?

jasmussen commented:
> or moved down near the socials?

Oh yeah, if you just move the button down at the end of the list of other social buttons, that's a great solution.

ramonjd commented:
Okay! Designs updated. It was good opportunity to replace the markup with WordPress components too.

<img width="711" alt="Screenshot 2025-03-12 at 11 17 18 am" src="https://github.com/user-attachments/assets/65d3baca-dc73-467f-96d8-7a08f474e0ad" />

ramonjd committed:
Refactor ShareSiteModal with improved input and sharing functionality

- Replace domain text with InputControl for better UX
- Add memoized share links with dynamic rendering
- Improve type definitions for share links
- Optimize performance with useCallback and useMemo
- Update styling for input and social share buttons

andrewserong reviewed (approved):
Looking good to me now!

<img width="684" alt="image" src="https://github.com/user-attachments/assets/92af41e9-5ea9-4c44-a8ff-bf3bab362c0b" />

If @jasmussen is happy with it, I reckon this'll be good to land 🚀

jasmussen commented:
Go for it!

jameskoster commented:
Not for now, but one other idea would be to do something a bit like the Block Inserter. I'm not sure we need to show the URL, and there has been a lot of feedback about that copy URL button.

<img width="737" alt="Screenshot 2025-03-12 at 12 51 41" src="https://github.com/user-attachments/assets/900fd058-5e7c-4bb7-b58c-8dae56446a3f" />

crisbusquets commented:
It works for me, @ramonjd and @andrewserong 🚢  Can we add tracking here? This way we'll be able to know if people use the social links, and which ones.

And thanks Joen and Jay for jumping in!

ramonjd commented:
> Can we add tracking here? This way we'll be able to know if people use the social links, and which ones.

I've found a couple of appropriate events that are already registered, and I think `calypso_sharing_buttons_share_button_click` would be the most appropriate. It has properties to record context so can be integrated in launchpad funnels if required.

ramonjd committed:
Enhance ShareSiteModal with task tracking and improved sharing functionality

- Integrate task parameter into ShareSiteModal for better context during sharing
- Update trackShareClick function to log events with task context
- Refactor share link handling to improve user interaction and analytics tracking

ramonjd commented:
`calypso_subscribers_share_site` registered and validated as working 👍🏻

andrewserong reviewed (approved):
> calypso_subscribers_share_site registered and validated as working 👍🏻

Yep, these events appear to be being sent correctly via the Network tab in DevTools to me! 🚀

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17294911

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @ramonjd for including a screenshot in the description! This is really helpful for our translators.
